index.js

```diff
@@ -11,29 +11,146 @@ const CLIENT_INFO = `${sdkPkg.name}/${sdkPkg.version}`;
+  }
+
+  return [...new Set(allWidths)].filter(e => e); // Removing duplicates
+};
+
+async function changeWindowDimensionAndWait(driver, width, height, resizeCount) {
+  try {
+    if (CDP_SUPPORT_SELENIUM && driver?.capabilities?.browserName === 'chrome') {
+      await driver?.executeCdpCommand('Emulation.setDeviceMetricsOverride', {
+        height,
+        width,
+        deviceScaleFactor: 1,
+        mobile: false
+      });
+    } else {
+      await driver.manage().window().setRect({ width, height });
+    }
+  } catch (e) {
+    log.warn(`Resizing using CDP failed, falling back to driver resize for width ${width}`, e);
+    await driver.manage().window().setRect({ width, height });
+  }
+
+  try {
+    await driver.wait(async () => {
+      /* istanbul ignore next: no instrumenting injected code */
+      await driver.executeScript('return window.resizeCount') === resizeCount;
+    }, 1000);
+  } catch (e) {
+    log.warn(`Error while window resize event for width ${width}`, e);
+  }
+}
+
+// Captures responsive DOM snapshots across different widths
+async function captureResponsiveDOM(driver, options) {
+  const widths = getWidthsForMultiDOM(options.widths || [], utils.percy?.widths);
+  const domSnapshots = [];
+  const windowSize = await driver.manage().window().getRect();
+  let currentWidth = windowSize.width; let currentHeight = windowSize.height;
+  let lastWindowWidth = currentWidth;
+  let resizeCount = 0;
+
+  // Setup the resizeCount listener if not present
+  /* istanbul ignore next: no instrumenting injected code */
+  await driver.executeScript(`
+      if (!window.resizeCount) {
+          let resizeTimeout = false;
+          window.addEventListener('resize', () => {
+              if (resizeTimeout) clearTimeout(resizeTimeout);
+              resizeTimeout = setTimeout(() => window.resizeCount++, 100);
+          });
+      }
+      window.resizeCount = 0;
+  `);
+
+  for (let width of widths) {
+    if (lastWindowWidth !== width) {
+      resizeCount++;
+      await changeWindowDimensionAndWait(driver, width, currentHeight, resizeCount);
+      lastWindowWidth = width;
+    }
+
+    if (process.env.RESPONSIVE_CAPTURE_SLEEP_TIME) {
+      await new Promise(resolve => setTimeout(resolve, parseInt(process.env.RESPONSIVE_CAPTURE_SLEEP_TIME)));
+    }
+
+    let domSnapshot = await captureSerializedDOM(driver, options);
+    domSnapshot.width = width;
+    domSnapshots.push(domSnapshot);
+  }
+
+  // Reset window size back to original dimensions
+  await changeWindowDimensionAndWait(driver, currentWidth, currentHeight, resizeCount + 1);
+  return domSnapshots;
+}
+
+async function captureSerializedDOM(driver, options) {
+  /* istanbul ignore next: no instrumenting injected code */
+  let { domSnapshot } = await driver.executeScript(options => ({
+    /* eslint-disable-next-line no-undef */
+    domSnapshot: PercyDOM.serialize(options)
+  }), options);
+  domSnapshot.cookies = await driver.manage().getCookies();
```

Comment:
instead of getting cookies everytime can we call it once and reuse that

Reply:
DOM change resize change if it changes cookies will also change.